core/shared/src/main/scala/cats/effect/CpuStarvationCheck.scala

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2020-2022 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats.effect
+
+import cats.effect.std.Console
+import cats.effect.unsafe.IORuntimeConfig
+import cats.syntax.all._
+
+private[effect] object CpuStarvationCheck {
+  def run(runtimeConfig: IORuntimeConfig): IO[Unit] =
+    IO.monotonic
+      .flatMap { now =>
+        IO.sleep(runtimeConfig.cpuStarvationCheckInterval) >> IO
+          .monotonic
+          .map(_ - now)
+          .flatMap { delta =>
+            Console[IO]
+              .errorln("[WARNING] your CPU threadpool is probably starving")
+              .whenA(delta > runtimeConfig.cpuStarvationCheckThreshold)
```

Comment:
Couple thoughts on this…

First, I think we should probably have the starvation threshold be a *multiple* of the interval. In other words, the check threshold should be a `Double` like `0.1` or `0.2` (representing 10% of 20%, respectively). Then the calculation itself would be something like this:

```suggestion
      .flatMap { now =>
        IO.sleep(runtimeConfig.cpuStarvationCheckInterval) >> IO
          .monotonic
          .map(_ - now)
          .flatMap { delta =>
            Console[IO]
              .errorln("[WARNING] Your CPU is probably starving. Consider reducing the granularity of your `delay`s or adding more `cede`s. This may also be a sign that you are unintentionally running blocking I/O operations (such as File or InetAddress) without the `blocking` combinator.")
              .whenA(delta >= runtimeConfig.cpuStarvationCheckInterval.toDouble * (1 + runtimeConfig.cpuStarvationCheckThreshold))
```

My second thought is that we should be more explanatory in the error message. I took a stab at it in the above. Probably needs line wrapping at the very least.

Reply:
Yep, thank you!